url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/DiskStorageManager.java

```diff
@@ -0,0 +1,423 @@
+			}
+		}
+	}
+
+	/**
+	 * save the index to file
+	 */
+	@Override
+	public void flush() {
+		try {
+			m_indexFile.seek(0l);
+
+			m_indexFile.writeInt(m_pageSize);
+			m_indexFile.writeInt(m_nextPage);
+
+			int id, page;
+			int count = m_emptyPages.size();
+
+			m_indexFile.writeInt(count);
+
+			Iterator<Integer> it = m_emptyPages.iterator();
+			while (it.hasNext()) {
+				page = it.next().intValue();
+				m_indexFile.writeInt(page);
+			}
+
+			count = m_pageIndex.size();
+			m_indexFile.writeInt(count);
+
+			Iterator<Map.Entry<Integer, DiskStorageManager.Entry>> it2 = m_pageIndex
+					.entrySet().iterator();
+
+			while (it2.hasNext()) {
+				Map.Entry<Integer, DiskStorageManager.Entry> me = it2.next();
+				id = me.getKey().intValue();
+				m_indexFile.writeInt(id);
+
+				Entry e = me.getValue();
+				count = e.m_length;
+				m_indexFile.writeInt(count);
+
+				count = e.m_pages.size();
+				m_indexFile.writeInt(count);
+
+				for (int cIndex = 0; cIndex < count; cIndex++) {
+					page = e.m_pages.get(cIndex).intValue();
+					m_indexFile.writeInt(page);
+				}
+			}
+		} catch (IOException ex) {
+			throw new IllegalStateException("Corrupted index file.");
+		}
+	}
+
+	/**
+	 * load all data of the id-th entry (data in e.m_pages)
+	 */
+	@Override
+	public byte[] loadByteArray(final int id) {
+		Entry e = m_pageIndex.get(new Integer(id));
+		if (e == null) {
+			throw new InvalidPageException(id);
+		}
+
+		int cNext = 0;
+		int cTotal = e.m_pages.size();
+
+		byte[] data = new byte[e.m_length];
+		int cIndex = 0;
+		int cLen;
+		int cRem = e.m_length;
+
+		do {
+			try {
+				//locate every page that used by every entry
+				m_dataFile.seek(e.m_pages.get(cNext).intValue() * m_pageSize);
+				int bytesread = m_dataFile.read(m_buffer);
+				if (bytesread != m_pageSize) {
+					throw new IllegalStateException("Corrupted data file.");
+				}
+			} catch (IOException ex) {
+				throw new IllegalStateException("Corrupted data file.");
+			}
+
+			cLen = cRem > m_pageSize ? m_pageSize : cRem;
+			System.arraycopy(m_buffer, 0, data, cIndex, cLen);
+
+			cIndex += cLen;
+			cRem -= cLen;
+			cNext++;
+		} while (cNext < cTotal);
+
+		return data;
+	}
+
+	/**
+	 * store data into the id-th entry
+	 */
+	@Override
+	public int storeByteArray(final int id, final byte[] data) {
+		if (id == NewPage) {
+			Entry e = new Entry();
+			e.m_length = data.length;
+
+			int cIndex = 0;
+			int cPage;
+			int cRem = data.length;
+			int cLen;
+
+			while (cRem > 0) {
+				if (!m_emptyPages.isEmpty()) {
+					Integer i = m_emptyPages.first();
+					m_emptyPages.remove(i);
+					cPage = i.intValue();
+				} else {//no empty page remained
+					cPage = m_nextPage;
+					m_nextPage++;
+				}
+
+				cLen = cRem > m_pageSize ? m_pageSize : cRem;
+				//copy to m_buffer step by step
+				System.arraycopy(data, cIndex, m_buffer, 0, cLen);
+
+				try {
+					m_dataFile.seek(cPage * m_pageSize);
+					m_dataFile.write(m_buffer);
+				} catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 379.